### PR TITLE
Adding bunch of unary foreach APIs

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47385 Added foreach_trunc, foreahc_reciprocal, foreach_sigmoid APIs
* #47384 Added foreach_frac API
* **#47383 Adding bunch of unary foreach APIs**

Differential Revision: [D24737050](https://our.internmc.facebook.com/intern/diff/D24737050)

Implementing several unary operators for _foreach_ APIs.
### Planned list of ops 
- [x]  abs
- [x]  acos
- [x]  asin
- [x]  atan
- [x]  ceil
- [x]  cos
- [x]  cosh
- [x]  erf
- [x]  erfc
- [x]  exp
- [x]  expm1
- [x]  floor
- [x]  log
- [x]  log10
- [x]  log1p
- [x]  log2
- [ ]  frac
- [x]  neg
- [ ]  reciprocal
- [x]  round
- [ ]  rsqrt
- [ ]  sigmoid
- [x]  sin
- [x]  sinh
- [x]  sqrt
- [x]  tan
- [x]  tanh
- [ ]  trunc
- [x]  lgamma
- [ ]  digamma
- [ ]  erfinv
- [ ]  sign
- [ ]  mvlgamma
- [ ]  clamp
- [ ]  clamp_min
- [ ]  clamp_max

### Perf results
```
----------------- OP:  sin  -----------------
  Median: 998.79 us
  300.84 us

----------------- OP:  abs  -----------------
  Median: 1.19 ms
  294.97 us

----------------- OP:  acos  -----------------
  Median: 982.30 us
  299.40 us

----------------- OP:  asin  -----------------
  Median: 1.16 ms
  298.09 us

----------------- OP:  atan  -----------------
  Median: 986.92 us
  295.64 us

----------------- OP:  ceil  -----------------
  Median: 1.17 ms
  297.25 us

----------------- OP:  cos  -----------------
  Median: 972.72 us
  294.41 us

----------------- OP:  cosh  -----------------
  Median: 1.17 ms
  294.97 us

----------------- OP:  erf  -----------------
  Median: 1.17 ms
  297.02 us

----------------- OP:  erfc  -----------------
  Median: 1.14 ms
  299.23 us

----------------- OP:  exp  -----------------
  Median: 1.15 ms
  298.79 us

----------------- OP:  expm1  -----------------
  Median: 1.17 ms
  291.79 us

----------------- OP:  floor  -----------------
  Median: 1.17 ms
  293.51 us

----------------- OP:  log  -----------------
  Median: 1.13 ms
  318.01 us

----------------- OP:  log10  -----------------
  Median: 987.17 us
  295.57 us

----------------- OP:  log1p  -----------------
  Median: 1.13 ms
  297.15 us

----------------- OP:  log2  -----------------
  Median: 974.21 us
  295.01 us

----------------- OP:  frac  -----------------
  Median: 1.15 ms
  296.01 us

----------------- OP:  neg  -----------------
  Median: 1.13 ms
  294.98 us

----------------- OP:  reciprocal  -----------------
  Median: 1.16 ms
  293.69 us

----------------- OP:  round  -----------------
  Median: 1.12 ms
  297.48 us

----------------- OP:  sigmoid  -----------------
  Median: 1.13 ms
  296.53 us

----------------- OP:  sin  -----------------
  Median: 991.02 us
  295.78 us

----------------- OP:  sinh  -----------------
  Median: 1.15 ms
  295.70 us

----------------- OP:  sqrt  -----------------
  Median: 1.17 ms
  297.75 us

----------------- OP:  tan  -----------------
  978.20 us
  297.99 us

----------------- OP:  tanh  -----------------
  Median: 967.84 us
  297.29 us

----------------- OP:  trunc  -----------------
  Median: 1.14 ms
  298.72 us

----------------- OP:  lgamma  -----------------
  Median: 1.14 ms
  317.53 us
```

### Script

```

import torch
import torch.optim as optim
import torch.nn as nn
import torchvision
import torch.utils.benchmark as benchmark_utils

inputs = [torch.rand(3, 200, 200, device="cuda") for _ in range(100)]

def main():
    for op in [
            "sin", "abs", "acos", "asin", "atan", "ceil",
            "cos", "cosh", "erf", "erfc",
            "exp", "expm1", "floor", "log",
            "log10", "log1p", "log2", "frac",
            "neg", "reciprocal", "round",
            "sigmoid", "sin", "sinh", "sqrt",
            "tan", "tanh", "trunc", "lgamma"
        ]:
        print("\n\n----------------- OP: ", op, " -----------------")
        stmt = "[torch.{op}(t) for t in inputs]"
        timer = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer)",
        )
        print(f"autorange:\n{timer.blocked_autorange()}\n\n")


        stmt = "torch._foreach_{op}(inputs)"
        timer_mta = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer_mta)",
        )
        print(f"autorange:\n{timer_mta.blocked_autorange()}\n\n")

if __name__ == "__main__":
    main()

```

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 1]
```diff
@@ -70,7 +110,75 @@ void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
...
+                                                                        \
+    foreach_unary_op_no_comlex_<NAME1>(tensors);                        \
+}
+
 FOREACH_UNARY_OP(exp, Exp);
 FOREACH_UNARY_OP(sqrt, Sqrt);
+FOREACH_UNARY_OP(acos, Acos);
+FOREACH_UNARY_OP(asin, Asin);
+FOREACH_UNARY_OP(atan, Atan);
+FOREACH_UNARY_OP(cos, Cos);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:144]: `cos` supports bfloat16 on cuda, for parity mta should also dispatch to bfloat16

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 2]
```diff
@@ -70,7 +110,75 @@ void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
...
+FOREACH_UNARY_OP(cosh, Cosh);
+FOREACH_UNARY_OP(log, Log);
+FOREACH_UNARY_OP(log10, Log10);
+FOREACH_UNARY_OP(log2, Log2);
+FOREACH_UNARY_OP(tan, Tan);
+FOREACH_UNARY_OP(tanh, Tanh);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:150]: tanh supports bfloat16

File: aten/src/ATen/native/ForeachOpsKernels.cpp [hunk 3]
```diff
@@ -149,20 +149,48 @@ void foreach_tensor_##OP##_scalarlist_slow_(TensorList input, TensorList tensors
...
+FOREACH_UNARY_OP(asin);
+FOREACH_UNARY_OP(atan);
+FOREACH_UNARY_OP(ceil);
+FOREACH_UNARY_OP(cos);
+FOREACH_UNARY_OP(cosh);
+FOREACH_UNARY_OP(erf);
+FOREACH_UNARY_OP(erfc);
+FOREACH_UNARY_OP(expm1);
+FOREACH_UNARY_OP(floor);
+FOREACH_UNARY_OP(log);
```
Comment (ngimel) [inline aten/src/ATen/native/ForeachOpsKernels.cpp:179]: log and friends support bfloat16

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 4]
```diff
@@ -70,7 +110,75 @@ void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
...
+                                                                        \
+    if (!can_use_fast_route(tensors)) {                                 \
+        return at::native::foreach_tensor_##NAME##_slow_(tensors);      \
+    }                                                                   \
+                                                                        \
+    foreach_unary_op_no_comlex_<NAME1>(tensors);                        \
+}
+
 FOREACH_UNARY_OP(exp, Exp);
 FOREACH_UNARY_OP(sqrt, Sqrt);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:140]: sqrt supports bfloat16

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 5]
```diff
@@ -70,7 +110,75 @@ void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
...
+FOREACH_UNARY_OP(cosh, Cosh);
+FOREACH_UNARY_OP(log, Log);
+FOREACH_UNARY_OP(log10, Log10);
+FOREACH_UNARY_OP(log2, Log2);
+FOREACH_UNARY_OP(tan, Tan);
+FOREACH_UNARY_OP(tanh, Tanh);
+FOREACH_UNARY_OP(sin, Sin);
+FOREACH_UNARY_OP(sinh, Sinh);
+
+FOREACH_UNARY_OP_NO_COMPLEX(abs, Abs);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:154]: abs (I think?) supports bfloat16

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 6]
```diff
@@ -70,7 +110,75 @@ void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
...
+FOREACH_UNARY_OP_NO_COMPLEX(ceil, Ceil);
+FOREACH_UNARY_OP_NO_COMPLEX(erf, Erf);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:156]: erf supports bfloat16

File: test/test_foreach.py [hunk 7]
```diff
@@ -150,12 +150,44 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
     # Unary ops
     #
     @dtypes(*[torch.float, torch.double, torch.complex64, torch.complex128])
```
Comment (ngimel) [inline test/test_foreach.py:152]: it would make sense to test against all floating types here and make sure that errors thrown are the same for slow and fast paths - that way we can track that bfloat16 support is on par.

**Review by ngimel — COMMENTED** [review]
This generally looks good, but due to intermediate state of our bfloat16 support there are some unary functions that generally support bfloat16, but you don't dispatch their foreach implementations to bfloat16. 
I've commented on some of them, but the list is not exhaustive I suspect, and it's better tracked with testing (mta and cuda loop implementation should throw the same errors)

**Comment by izdeby** [comment]
@ngimel, resolved all the comments. All unary ops support half (in CUDA case) and there is a map of bfloat16/complex64/complex128 support in tests

**Review by ngimel — APPROVED** [review]
Great! I hope we can get rid of dispatches without bfloat16 soon, as regular ops get better bfloat16 support.

**Comment by jeffdaily** [comment]
@ngimel @izdeby this PR appears to have broken ROCm CI -- can you revert?
